### PR TITLE
Package Declarations4

### PR DESCRIPTION
**Pull request by knewbury01**
## Description

_please enter the description of your change here_

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [x] Queries have been added for the following rules:
     - RULE-8-2
     - RULE-8-3
     - RULE-8-4
     - RULE-8-6
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [ ] Yes
  - [x] No

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [ ] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### Reviewer 

 - [ ] Have all the relevant rule package description files been checked in?
 - [ ] Have you verified that the metadata properties of each new query is set appropriately?
 - [ ] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [ ] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [ ] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [ ] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### CODE REVIEW

**Comment by knewbury01** [comment]
# Notes RULE-8-6

this is the same as M3-2-4 but is using a [slightly different previously agreed upon](https://github.com/github/codeql-coding-standards/pull/39#discussion_r945615958) defn of external identifiers, therefore was not directly imported

**Comment by knewbury01** [comment]
# Notes RULE-8-3

this is another _almost_ import (this time of `M3-2-1`) but with a few minor differences, can just do as import if the small differences dont seem to make a difference overall

File: c/misra/src/rules/RULE-8-3/DeclarationsOfAnObjectSameNameAndType.ql [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/**
+ * @id c/misra/declarations-of-an-object-same-name-and-type
+ * @name RULE-8-3: All declarations of an object shall use the same names and type qualifiers
```
Comment (lcartey) [inline c/misra/src/rules/RULE-8-3/DeclarationsOfAnObjectSameNameAndType.ql:3]: This looks very similar to M3-2-1 `DeclarationsOfAnObjectShallHaveCompatibleTypes.ql` - can we have a shared implementation between the two?

File: cpp/common/src/codingstandards/cpp/Compatible.qll [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import cpp
+
+predicate typesCompatible(Type t1, Type t2) {
```
Comment (lcartey) [inline cpp/common/src/codingstandards/cpp/Compatible.qll:3]: M3-2-1 uses `areCompatible` for this, but it's a barebones predicate. Can we combine these two predicates together to share across both sets of queries?

**Review by lcartey — CHANGES_REQUESTED** [review]
Added some initial review comments.

File: c/misra/src/rules/RULE-8-3/DeclarationsOfAnObjectSameNameAndType.ql (hunk 1 above)
Comment (knewbury01) [reply c/misra/src/rules/RULE-8-3/DeclarationsOfAnObjectSameNameAndType.ql:3]: I dont see why not, one specific thing is that C RULE-8-3 specifically disallows the case where the basic types are the same (but typenames are different) and in M3-2-1  we were allowing that, but I wasnt sure if that was a purposeful choice

Comment (knewbury01) [reply c/misra/src/rules/RULE-8-3/DeclarationsOfAnObjectSameNameAndType.ql:3]: I mean we can certainly change the strictness of  M3-2-1, I guess I just wanted the second opinion before doing that :)

File: c/misra/src/rules/RULE-8-6/IdentifierWithExternalLinkageOneDefinition.ql [hunk 3]
```diff
@@ -0,0 +1,38 @@
+/**
+ * @id c/misra/identifier-with-external-linkage-one-definition
+ * @name RULE-8-6: An identifier with external linkage shall have exactly one external definition
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/RULE-8-6/IdentifierWithExternalLinkageOneDefinition.ql:3]: sure thing, will apply

File: c/misra/src/rules/RULE-8-3/DeclarationsOfAnObjectSameNameAndType.ql (hunk 1 above)
Comment (knewbury01) [reply c/misra/src/rules/RULE-8-3/DeclarationsOfAnObjectSameNameAndType.ql:3]: discussed with @lcartey in meeting, we actually perhaps do not want to change the strictness in this case

File: c/misra/src/rules/RULE-8-3/Compatible.qll [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import cpp
+
+/*
+ * This is a copy of the `arithTypesMatch` predicate from the standard set of
+ * queries as of the `codeql-cli/2.9.4` tag in `github/codeql`.
+ */
+
+pragma[inline]
+predicate arithTypesMatch(Type t1, Type t2) {
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/RULE-8-3/Compatible.qll]: actually this comment is deprecated, I changed that implementation anyhow to something more correct

File: c/misra/src/rules/RULE-8-6/IdentifierWithExternalLinkageOneDefinition.ql (hunk 3 above)
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/RULE-8-6/IdentifierWithExternalLinkageOneDefinition.ql]: change applied!

File: c/misra/src/rules/RULE-8-2/FunctionTypesNotInPrototypeForm.ql [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    f.getAParameter() instanceof UnnamedParameter and
+    msg = "Function " + f + " declares parameter that is unnamed."
+    or
+    hasZeroParamDecl(f) and
+    msg = "Function " + f + " does not specifiy void for no parameters present."
+    or
+    exists(Parameter p |
+      p.getFunction() = f and
+      not p.getFile() = f.getFile() and
+      msg = "Function " + f + " declares parameter in unsupported declaration list."
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/RULE-8-2/FunctionTypesNotInPrototypeForm.ql:53]: yep, inline comment added!

File: c/misra/src/rules/RULE-8-2/FunctionTypesNotInPrototypeForm.ql [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    )
+  ) and
+  not f.isInMacroExpansion()
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/RULE-8-2/FunctionTypesNotInPrototypeForm.ql:56]: hm, I have no idea what I was thinking anymore haha oops, maybe was leftover from some debugging? makes sense to remove!

File: cpp/common/src/codingstandards/cpp/Compatible.qll (hunk 2 above)
Comment (knewbury01) [reply cpp/common/src/codingstandards/cpp/Compatible.qll:3]: see comment above for why not

**Review by lcartey — APPROVED** [review]
Merging!

**Pull request merged**